File: backend/app/services/sms_scheduler.py
"""
APScheduler-based SMS task runner.
Loads active tasks from DB and fires them on schedule.
"""
import logging
from datetime import datetime
from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.cron import CronTrigger
from apscheduler.triggers.date import DateTrigger
from sqlalchemy.orm import Session

from app.core.database import SessionLocal
from app.models.sms import SmsScheduledTask, SmsMessage, TaskStatus, SmsStatus, SmsDirection
from app.models.modem import Modem
from app.services import modem_manager

logger = logging.getLogger(__name__)
scheduler = AsyncIOScheduler()


def start():
    scheduler.start()
    scheduler.add_job(reload_tasks, "interval", seconds=60, id="task_reloader", replace_existing=True)
    reload_tasks()
    logger.info("SMS scheduler started")


def stop():
    scheduler.shutdown()


def reload_tasks():
    """Sync APScheduler jobs with DB tasks."""
    db = SessionLocal()
    try:
        tasks = db.query(SmsScheduledTask).filter(SmsScheduledTask.status == TaskStatus.ACTIVE).all()
        active_ids = set()
        for task in tasks:
            job_id = f"sms_task_{task.id}"
            active_ids.add(job_id)
            if scheduler.get_job(job_id):
                continue
            _schedule_task(task, db)
        # Remove jobs for deleted/paused tasks
        for job in scheduler.get_jobs():
            if job.id.startswith("sms_task_") and job.id not in active_ids:
                scheduler.remove_job(job.id)
    finally:
        db.close()


def _schedule_task(task: SmsScheduledTask, db: Session):
    job_id = f"sms_task_{task.id}"
    if task.cron_expression:
        trigger = CronTrigger.from_crontab(task.cron_expression)
    elif task.send_once_at:
        trigger = DateTrigger(run_date=task.send_once_at)
    else:
        logger.warning(f"Task {task.id} has no schedule, skipping")
        return

    job = scheduler.add_job(
        execute_task,
        trigger=trigger,
        args=[task.id],
        id=job_id,
        replace_existing=True,
    )
    if job.next_run_time:
        task.next_run_at = job.next_run_time.replace(tzinfo=None)
        db.commit()
    logger.info(f"Scheduled task {task.id} ({task.name})")


async def execute_task(task_id: int):
    db = SessionLocal()
    try:
        task = db.query(SmsScheduledTask).filter(SmsScheduledTask.id == task_id).first()
        if not task or task.status != TaskStatus.ACTIVE:
            return

        modem = db.query(Modem).filter(Modem.id == task.modem_id).first()
        if not modem or not modem.mm_object_path:
            logger.error(f"Task {task_id}: modem not found")
            return

        import re
        match = re.search(r"/Modem/(\d+)$", modem.mm_object_path)
        if not match:
            return
        mm_index = match.group(1)

        recipients = task.recipients if isinstance(task.recipients, list) else []
        for phone_number in recipients:
            success, message = modem_manager.send_sms(mm_index, phone_number, task.content)
            sms = SmsMessage(
                modem_id=modem.id,
                direction=SmsDirection.OUTBOUND,
                phone_number=phone_number,
                content=task.content,
                status=SmsStatus.SENT if success else SmsStatus.FAILED,
                error_message=None if success else message,
                sent_at=datetime.utcnow() if success else None,
                scheduled_task_id=task.id,
            )
            db.add(sms)

        task.last_run_at = datetime.utcnow()
        task.run_count = (task.run_count or 0) + 1
        if task.send_once_at:
            task.status = TaskStatus.COMPLETED
        db.commit()
        logger.info(f"Task {task_id} executed, sent to {len(recipients)} recipients")
    except Exception as e:
        logger.exception(f"Task {task_id} failed: {e}")
        db.rollback()
    finally:
        db.close()

File: backend-go/services/sms_scheduler.go
package services

import (
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"time"

	"simnexus-go/database"
	"simnexus-go/models"

	"github.com/robfig/cron/v3"
)

var (
	cronRunner *cron.Cron                 // cron 调度器，使用 UTC 时区
	jobIDs     = map[uint]cron.EntryID{}  // task.ID → cron entry ID，用于取消定期任务
	onceTimers = map[uint]*time.Timer{}   // task.ID → time.Timer，用于取消一次性任务
	schedMu    sync.Mutex                 // 保护 jobIDs 和 onceTimers 的并发访问
)

// StartScheduler launches the cron runner and the 60s reload loop.
func StartScheduler() {
	cronRunner = cron.New(cron.WithLocation(time.UTC))
	cronRunner.Start()
	ReloadTasks()
	go func() {
		ticker := time.NewTicker(60 * time.Second)
		for range ticker.C {
			ReloadTasks()
		}
	}()
	slog.Info("SMS scheduler started")
}

// ReloadTasks syncs cron/one-shot jobs with active DB tasks.
func ReloadTasks() {
	schedMu.Lock()
	defer schedMu.Unlock()

	var tasks []models.SmsScheduledTask
	database.DB.Where("status = ?", models.TaskActive).Find(&tasks)
	active := map[uint]bool{}
	now := time.Now().UTC()

	for _, task := range tasks {
		active[task.ID] = true
		if _, ok := jobIDs[task.ID]; ok {
			continue
		}
		if _, ok := onceTimers[task.ID]; ok {
			continue
		}
		// skip past-due one-shot tasks
		if task.SendOnceAt != nil && !task.SendOnceAt.After(now) {
			continue
		}
		scheduleLocked(task)
	}

	// remove jobs for deleted/paused tasks
	for id, entry := range jobIDs {
		if !active[id] {
			cronRunner.Remove(entry)
			delete(jobIDs, id)
		}
	}
	for id, t := range onceTimers {
		if !active[id] {
			t.Stop()
			delete(onceTimers, id)
		}
	}
}

// ScheduleTask registers a single task (used right after creation).
func ScheduleTask(task models.SmsScheduledTask) {
	schedMu.Lock()
	defer schedMu.Unlock()
	scheduleLocked(task)
}

// scheduleLocked 在持有 schedMu 锁的情况下注册单个任务，需调用方加锁。
// cron 任务注册到 cronRunner，一次性任务使用 time.AfterFunc。
func scheduleLocked(task models.SmsScheduledTask) {
	id := task.ID
	if task.CronExpression != nil && strings.TrimSpace(*task.CronExpression) != "" {
		entry, err := cronRunner.AddFunc(strings.TrimSpace(*task.CronExpression), func() {
			ExecuteTask(id)
		})
		if err != nil {
			slog.Error("invalid cron expression", "task_id", id, "err", err)
			return
		}
		jobIDs[id] = entry
		if next := cronRunner.Entry(entry).Next; !next.IsZero() {
			nx := next.UTC()
			database.DB.Model(&models.SmsScheduledTask{}).Where("id = ?", id).Update("next_run_at", nx)
		}
		slog.Info("scheduled cron task", "task_id", id, "name", task.Name)
	} else if task.SendOnceAt != nil {
		delay := time.Until(*task.SendOnceAt)
		if delay < 0 {
			delay = 0
		}
		onceTimers[id] = time.AfterFunc(delay, func() {
			ExecuteTask(id)
			schedMu.Lock()
			delete(onceTimers, id)
			schedMu.Unlock()
		})
		database.DB.Model(&models.SmsScheduledTask{}).Where("id = ?", id).Update("next_run_at", task.SendOnceAt.UTC())
		slog.Info("scheduled one-shot task", "task_id", id, "name", task.Name, "fire_at", task.SendOnceAt)
	}
}

// RemoveTask cancels any scheduled job for the task.
func RemoveTask(taskID uint) {
	schedMu.Lock()
	defer schedMu.Unlock()
	if entry, ok := jobIDs[taskID]; ok {
		cronRunner.Remove(entry)
		delete(jobIDs, taskID)
	}
	if t, ok := onceTimers[taskID]; ok {
		t.Stop()
		delete(onceTimers, taskID)
	}
}

// ExecuteTask fires a task: sends to all recipients and updates status.
func ExecuteTask(taskID uint) {
	db := database.DB
	var task models.SmsScheduledTask
	if db.First(&task, taskID).Error != nil || task.Status != models.TaskActive {
		return
	}
	var modem models.Modem
	if db.First(&modem, task.ModemID).Error != nil || modem.MmObjectPath == "" {
		slog.Warn("task modem not found", "task_id", taskID)
		return
	}

	obj := modem.MmObjectPath
	isZte := strings.HasPrefix(obj, "zte:")
	mmIndex := ""
	if !isZte {
		m := reModemIdx.FindStringSubmatch(obj)
		if m == nil {
			return
		}
		mmIndex = m[1]
	}

	recipients := []string(task.Recipients)
	failCount := 0
	for _, phone := range recipients {
		var success bool
		var message string
		if isZte {
			success = ZteSendSMS(phone, task.Content)
			if !success {
				message = "ZTE send failed"
			}
		} else {
			success, message = SendSMS(mmIndex, phone, task.Content)
		}
		now := time.Now()
		sms := models.SmsMessage{
			ModemID:         modem.ID,
			Direction:       models.SmsOutbound,
			PhoneNumber:     phone,
			Content:         task.Content,
			Status:          models.SmsSent,
			ScheduledTaskID: &task.ID,
			CreatedByID:     task.CreatedByID,
		}
		if success {
			sms.SentAt = &now
		} else {
			sms.Status = models.SmsFailed
			em := message
			sms.ErrorMessage = &em
			failCount++
		}
		db.Create(&sms)
	}

	if failCount > 0 {
		label := modem.Alias
		if label == "" {
			label = modem.Model
		}
		if label == "" {
			label = fmt.Sprintf("设备#%d", modem.ID)
		}
		body := fmt.Sprintf("任务「%s」[%s] 有 %d/%d 条短信发送失败", task.Name, label, failCount, len(recipients))
		if task.CreatedByID != nil {
			var creator models.User
			if db.First(&creator, *task.CreatedByID).Error == nil && creator.IsAdmin() {
				Push("task_failed", "定时任务失败", body, "admin", nil)
			} else {
				Push("task_failed", "定时任务失败", body, "user", task.CreatedByID)
			}
		} else {
			Push("task_failed", "定时任务失败", body, "admin", nil)
		}
	}

	now := time.Now()
	task.LastRunAt = &now
	task.RunCount++
	if task.SendOnceAt != nil {
		if failCount == len(recipients) {
			task.Status = models.TaskFailed
		} else {
			task.Status = models.TaskCompleted
		}
	}
	db.Save(&task)
	slog.Info("task executed", "task_id", taskID, "recipients", len(recipients), "failed", failCount)
}
